core/handler/device/store.go

```diff
@@ -23,11 +24,13 @@ type Store interface {
 	DownlinkQueue(appID, devID string) (DownlinkQueue, error)
 	Set(new *Device, properties ...string) (err error)
 	Delete(appID, devID string) error
+	AttrWhitelist() map[string]bool
 }
 
 const defaultRedisPrefix = "handler"
 const redisDevicePrefix = "device"
 const redisDownlinkQueuePrefix = "downlink"
+const redisUpStreamMetadataKeyWhitelist = ":attr-whitelist"
```

Comment:
1. [Yes](https://github.com/TheThingsNetwork/ttn/blob/master/api/validation.go)
2. Too geeky; there's no reason for in this case, nothing Redis optimises here for example
3. Yep. For me, "standard" is "builtin" as it clearly indicates that it's standard or builtin. Common can be anything people commonly use, and everything here is already "ttn" so that may confuse. From an outside perspective, all these attributes are "ttn attributes"
4. Get the attributes of the device, send the standard ones anyway, sort the other ones alphabetically and select the top n from them.